connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java

```diff
@@ -294,24 +304,45 @@ private PluginScanResult scanPluginPath(
 
         Collection<PluginDesc<T>> result = new ArrayList<>();
         for (Class<? extends T> plugin : plugins) {
-            if (PluginUtils.isConcrete(plugin)) {
-                // Temporary workaround until all the plugins are versioned.
-                if (Connector.class.isAssignableFrom(plugin)) {
-                    result.add(
-                            new PluginDesc<>(
-                                    plugin,
-                                    ((Connector) plugin.newInstance()).version(),
-                                    loader
-                            )
-                    );
-                } else {
-                    result.add(new PluginDesc<>(plugin, "undefined", loader));
-                }
-            }
+            addPluginDesc(loader, result, plugin);
+        }
+        return result;
+    }
+
+    private <T> Collection<PluginDesc<T>> getServiceLoaderPluginDesc(
+        Class<T> klass,
+        ClassLoader loader
+    ) throws IllegalAccessException, InstantiationException {
+        Iterator<T> pluginsIterator = ServiceLoader.load(klass, loader).iterator();
+
+        Collection<PluginDesc<T>> result = new ArrayList<>();
+        while (pluginsIterator.hasNext()) {
+            addPluginDesc(loader, result, (Class<? extends T>) pluginsIterator.next());
```

Comment:
Good catch @rhauch on the ServiceLoader returning an implementation. Having said that, do you think we should cache these implementations instead of creating an instance ourselves? I don't personally see a benefit to using the same instance the one returned by the ServiceLoader.

Reply:
The ServiceLoader is usually just used directly to load all instances of a particular extension point, but that's not really what we do. And really there's nothing special about _how_ ServiceLoader instantiates the class: it really just does a new instance on the class.

Reply:
BTW, we should have a test case that loads a test implementation of `RestExtension` using Plugins.

Reply:
@rhauch yes, I will be adding more tests. Once we all agree on the public interfaces, I will update the KIP and refine this PR